### PR TITLE
The strip lights up!

### PR DESCRIPTION
**Pull request by raldi**

### CODE REVIEW

**Comment by raldi** [comment]
I've got a strand of lights sitting next to me that this code is successfully illuminating!

But it's still a WIP because it's missing one very important feature. 

Let's say our model has two edges, e0 and e1, each with 50 pixels. We have a physical strand plugged into the sACN device that's 100 pixels long, configured to correspond to Universe 1, and it's going to be attached to the real-life model such that its first 50 pixels are e0 and the next 50 are e1. We'd configure them as follows:

e0 universe=1 strandOffset=0
e1 universe=1 strandOffset=50

The problem is, the code as it stands now is just ignoring the strandOffset, because each edge's points are just mapped directly to the StreamingACNDatagram's data buffer. I think we need to somehow create one StreamingACNDatagram for both edges to use, and somehow map both of their points, concatenated, to its data buffer. I'll start investigating that, but if Mark or anyone else wants to call out the answer, that would be okay too.

**Comment by raldi** [comment]
Oh, also there's this in my logs:

[LX 2022/02/10 14:29:37] Calls to DatagramSocket.send() appear to be unexpectedly blocking, you may be sending to an unresolvable address or network queues may be saturated. If you are on Linux/Raspberry-Pi, consult the following URL for guidance on relevant kernel parameters: https://github.com/heronarts/LXStudio/wiki/Raspberry-Pi

File: LXStudio-IDE/src/main/java/titanicsend/output/TESacnOutput.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      }
+      assert edgeEntry.universeNum == currentUniverseNum;
+      assert edgeEntry.strandOffset == currentStrandOffset;
+      String edgeSummary = "[" + currentStrandOffset + ":Edge_" + edgeEntry.edge.id() + "=" + edgeLength + "] ";
+      logString.append(edgeSummary);
+      currentStrandOffset += edgeLength;
+      this.deviceLengths.put(currentUniverseNum, currentStrandOffset);
+
+      // FIXME: We're not doing anything with the strand offset.
+      this.outputDevice = new StreamingACNDatagram(lx, edgeEntry.edge, currentUniverseNum);
```
Comment (tcahill) [inline LXStudio-IDE/src/main/java/titanicsend/output/TESacnOutput.java]: I believe you'll want to use the `StreamingACNDatagram(LX lx, int[] indexBuffer, int universeNumber)` constructor instead of the one that passes in the whole model. `indexBuffer` here is an array of the `index` of each `LXPoint` that should be controlled by the output. The constructor you have here actually uses [`LXModel.toIndexBuffer`](https://github.com/heronarts/LX/blob/master/src/main/java/heronarts/lx/model/LXModel.java#L703) under the hood.

Comment (tcahill) [reply LXStudio-IDE/src/main/java/titanicsend/output/TESacnOutput.java]: That way, you can combine the indices of each edge covered by this output into a single `indexBuffer`.

Comment (raldi) [reply LXStudio-IDE/src/main/java/titanicsend/output/TESacnOutput.java]: That all sounds right, but here's what's weird: the code above is only called once, at initialization time, and then somehow LX knows to keep pulling the model data out and sending datagrams with it, even though I'm not touching any of these data structures in any of TE's own loop code. I'm guessing there's some indirection going on, but I haven't had time to trace it all out with the debugger.

To put it another way, after we build our own concatenated indexBuffer, what do we do with it next? And is this done at initialization time or on every loop?

Comment (tcahill) [reply LXStudio-IDE/src/main/java/titanicsend/output/TESacnOutput.java]: Here's how I understand it. For simplicity, imagine we have a strip spanning 2 edges with a total of 4 pixels with the indexes `{0, 1, 4, 5}` (where by index, I mean `LXPoint.index` - an integer unique to each pixel in the top-level model that points into a global `colors` buffer). Now we create an sACN output like this: `StreamingACNDatagram(lx, new int[]{0, 1, 4, 5})`. The LX engine calls [`LXDatagram.onSend`](https://github.com/heronarts/LX/blob/master/src/main/java/heronarts/lx/output/LXDatagram.java#L198) on every loop, which in turn calls [`LXBufferOutput.updateDataBuffer`](https://github.com/heronarts/LX/blob/master/src/main/java/heronarts/lx/output/LXBufferOutput.java#L110) to insert the current colors for the pixels controlled by our output into the buffer that stores the data we send over the wire. The fact that this code is only called once is fine as long as our LXPoint indices don't change later.

To maybe clear up some confusion that I first had when reading the code, `StreamingACNDatagram` is not just a single static datagram - it's actually an output that continuously transmits data in the form of sACN datagrams.

**Comment by mcslee** [comment]
Hey guys - just confirming that @tcahill has got it right here. When you're using a hardcoded model, LXOutput objects should be set up once, at initialization time. Typically immediately after you build your model. When you add one of those LXOutput classes (e.g. a StreamingACNDatagram) to the engine, it will take the updated color buffer from the animation engine on every frame and send it. There are also some options on there to do things like gamma correction.

The animation engine works on a big flat 1-d array, which is what colors[] is that you compute inside a pattern. Note that there's not just *one* colors[] array, the mixer engine blends multiple of them together, but they're all indexed the same way, based upon the LXModel. LXPoint.index tells you which address in the raster corresponds to this point in the model.

The basic output classes do not deal with splitting packets, overflowing across packet boundaries, etc. So if you have a load of fixtures wired up in series that are going to overlap many universes, you will need to do the array math chunking yourself to set up the appropriate index buffers for each packet.

Here's pseudocode, don't have the correct model 
```java
LXModel[] componentsInSerial = /* somehow you've gotten the set of model components all on an output together */;
List<int> fullIndexBuffer = new ArrayList<int>();
for (LXModel component : componentsInSerial) {
  for (LXPoint p : component.points) {
    fullIndexBuffer.add(p.index);
  }
}
// Now, assuming these are streamingACN packets with size 512 RGB so 170 pixels per packet
// split fullIndexBuffer into chunks of 170
for (170-chunk-index-buffer) {
  new StreamingACNDatagram(lx, chunkIndexBuffer, universeNum)
  universeNum++
}
```

Let me know if specific questions come up. The chunking code can be a bit annoying as you do need to know all of the pixels that want to be in a particular universe packet before constructing it. The UX-based fixture system in LX does that, here is some code you might want to reference: https://github.com/heronarts/LX/blob/master/src/main/java/heronarts/lx/structure/LXStructure.java#L60

The LXStructure.Output class generates outputs based upon fixture definitions from the UX. Though this is dealing with a *lot* more possible complexity than you have here, so it may be overkill.

In your case, you know upfront how you're wiring your sculpture. I think if you just define some static arrays that specify which edges are wired in serial with other edges, then you can just slam all the point indices together into a big ordered 1-D array, and then chunk it.

**Comment by raldi** [comment]
Thanks! This PR actually does all the hard work of getting the linear representations in order; by the time activate() returns, they're all sorted correctly and we've asserted there are no gaps (gap tolerance will be a feature of the next version 😉) .. I just didn't know about the indirection going on; I thought we were passing the datagram an array of constant values, not an array of references to values that change on each frame. Will set things up later today, try it out with the hardware, and post here about how it goes.

**Comment by mcslee** [comment]
> Oh, also there's this in my logs:
> 
> [LX 2022/02/10 14:29:37] Calls to DatagramSocket.send() appear to be unexpectedly blocking, you may be sending to an unresolvable address or network queues may be saturated. If you are on Linux/Raspberry-Pi, consult the following URL for guidance on relevant kernel parameters: https://github.com/heronarts/LXStudio/wiki/Raspberry-Pi

Sorry forgot to respond to this. Safe to ignore, I need to tweak that code to be less sensitive. On RPi there is an issue where if an address isn't resolving it will block up into the application layer for a really long time, pretty much bringing the whole program to a halt. So I put this message in there for people to discover the settings they can tweak on RPi to stop such aggressive blocking on network address issues.

Irrelevant on other platforms - but the first couple Datagram send calls are still usually slow, probably due to Java lazy-loading/initializing something or other. And the way I implemented this is firing that warning too aggressively. It's on my TODO list to wait a while after program startup before issuing this warning, to make sure it's really persistently happening. Sorry for the confusion.

**Comment by raldi** [comment]
Multiple edges work now, and panels too.
